### DISCUSSION TITLE
Bypassing dynamic shape limits in XLA for expanding N-body spatial hashing (String-Star Manifold v2.0)

### DISCUSSION THREAD

**Discussion by Rupayan52** [jax-ml/jax#37280] [Show and tell]
Hey JAX community,

I wanted to share a recent architectural follow-up from my research on the **String-Star Manifold**, an open-source cosmological engine built entirely in JAX to run on TPUs. 

We recently pushed the v2.0 update, which transitioned the simulation from a static spatial grid to a dynamically expanding universe (implementing the FLRW metric/Hubble Flow). 
Repository link: https://github.com/Rupayan52/String-Star-Manifold
DOI [10.5281/zenodo.19822536](https://doi.org/10.5281/zenodo.19822536)

### The XLA Challenge
Simulating metric expansion in an $O(N \log N)$ N-body spatial hash usually requires the coordinate grid to physically grow. However, dynamically resizing array structures or grid cell counts at each epoch immediately triggers dynamic shape recompilation errors in XLA, crashing the TPU pipeline.

### The Solution: "Stretchy" Spatial Hashing
To maintain compilation efficiency and keep array shapes perfectly static, I implemented a custom `compute_macro_nodes` function using `@jax.jit` and `jax.ops.segment_sum`.

Instead of changing the array sizes, the simulation multiplies the physical boundaries (`grid_min`, `grid_max`, `grid_res`) by the cosmic Scale Factor $a(t)$ at each timestep. The JAX arrays remain static in memory, but the "mathematical volume" they represent stretches. This allows the engine to aggregate distant macro-nodes with zero loss of computational efficiency or XLA recompilations.

**Core Implementation:**
```python
@jax.jit
def compute_macro_nodes(pos_array, effective_masses, scale_factor, base_grid_bounds):
    # Dynamically stretch the bounds inside the compiled function
    current_bounds = base_grid_bounds * scale_factor
    
    # Standard spatial hashing against the stretched bounds
    # ... (clip and hash positions) ...
    
    # Aggregate masses into macro-nodes statically
    cell_masses = jax.ops.segment_sum(effective_masses, cell_hashes, num_segments=TOTAL_CELLS)
    
    return cell_masses
